src/loop_rerolling.py

```diff
@@ -0,0 +1,138 @@
+from typing import Dict, List, Optional, Tuple, Union
+
+from .flow_graph import (
+    BasicNode,
+    ConditionalNode,
+    FlowGraph,
+    Node,
+    compute_relations,
+)
+from .parse_instruction import Instruction
+
+
+def replace_node_references(
+    flow_graph: FlowGraph, replace_this: Node, with_this: Node
+) -> None:
+    for node_to_modify in flow_graph.nodes:
+        node_to_modify.replace_any_children(replace_this, with_this)
+
+
+def remove_node(flow_graph: FlowGraph, to_delete: Node, new_child: Node) -> None:
+    flow_graph.nodes.remove(to_delete)
+    replace_node_references(flow_graph, to_delete, new_child)
+
+
+def replace_node(flow_graph: FlowGraph, replace_this: Node, with_this: Node) -> None:
+    replacement_index = flow_graph.nodes.index(replace_this)
+    flow_graph.nodes[replacement_index] = with_this
+    replace_node_references(flow_graph, replace_this, with_this)
+
+
+PatternGraph = Dict[int, Union[int, Tuple[int, int]]]
+
+IDO_O2_SIMPLE_LOOP: PatternGraph = {
+    0: (1, 7),
+    1: (2, 5),
+    2: 3,
+    3: (4, 3),
+    4: (5, 7),
+    5: 6,
+    6: (7, 6),
+}
+
+
+def detect_pattern(
+    pattern: PatternGraph, flow_graph: FlowGraph, start: Node
+) -> Optional[Tuple[Node, ...]]:
+    indices = [node.block.index for node in flow_graph.nodes]
+    assert sorted(indices) == indices, "FlowGraphs should be sorted"
+
+    offset = start.block.index
+    for label in pattern.keys():
+        try:
+            node = flow_graph.nodes[label + offset]
+            target = pattern[label]
+            if isinstance(target, int):
+                if not (
+                    isinstance(node, BasicNode)
+                    and node.successor is flow_graph.nodes[offset + target]
+                ):
+                    return None
+            else:
+                (fallthrough, conditional) = target
+                if not (
+                    isinstance(node, ConditionalNode)
+                    and node.conditional_edge is flow_graph.nodes[offset + conditional]
+                    and node.fallthrough_edge is flow_graph.nodes[offset + fallthrough]
+                ):
+                    return None
+        except IndexError:
+            return None
+
+    # TODO: Check that the subgraph is self-contained except for the entry
+    # and exit nodes, which themselves should be somehow "designated" using an enum.
```

Comment:
I think the following check is sufficient, without requiring you to explicitly mark an exit node in the pattern? (I think an exit node is any node that is not in `pattern.keys()` but is in `pattern.values()`? And the start node is `0`?) 
I think it can be up to the caller to check `start.parents`: I don't think the start's parents matter in this case.

```python
# pattern_nodes could be the `tuple(flow_graph.nodes[i] for i in sorted(all_nodes_in_pattern))` return value I suggest later
for node in pattern_nodes:
    if node is not start and not all(p in pattern_nodes for p in node.parents):
        return None
```